service/history/workflow/delete_manager.go

```diff
@@ -0,0 +1,248 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination delete_manager_mock.go
+
+package workflow
+
+import (
+	"context"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/worker/archiver"
+)
+
+type (
+	DeleteManager interface {
+		DeleteWorkflowExecution(namespaceID namespace.ID, we commonpb.WorkflowExecution, weCtx Context, ms MutableState, sourceTaskVersion int64) error
+		DeleteWorkflowExecutionByRetention(namespaceID namespace.ID, we commonpb.WorkflowExecution, weCtx Context, ms MutableState, sourceTaskVersion int64) error
+	}
+
+	DeleteManagerImpl struct {
+		shard          shard.Context
+		historyCache   Cache
+		config         *configs.Config
+		metricsClient  metrics.Client
+		archivalClient archiver.Client
+	}
+)
+
+var _ DeleteManager = (*DeleteManagerImpl)(nil)
+
+func NewDeleteManager(
+	shard shard.Context,
+	cache Cache,
+	config *configs.Config,
+	archiverClient archiver.Client,
+) *DeleteManagerImpl {
+	deleteManager := &DeleteManagerImpl{
+		shard:          shard,
+		historyCache:   cache,
+		metricsClient:  shard.GetMetricsClient(),
+		config:         config,
+		archivalClient: archiverClient,
+	}
+
+	return deleteManager
+}
+
+func (m *DeleteManagerImpl) DeleteWorkflowExecution(
```

Comment:
so the difference from DeleteWorkflowExecutionByRetention is this method allow delete running workflow? Why do we need to delete running workflow? Isn't that we require workflow to be closed before you can delete them?

Reply:
Because of cherry-picking I missed check for running workflow. Restored.